File: integration_tests/java_client_tests_reactor_real/client-app/src/test/java/dev/graphlink/test/CycleTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.types.Post;
import dev.graphlink.test.generated.types.User;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class CycleTest {

    private GraphLinkClient client;

    @BeforeEach
    void setUp() {
        client = newRealClient();
    }

    // ── User → posts list → post.author (truncated, no posts) ─────────────────

    @Nested
    class UserWithPosts {

        @Test
        void user_posts_isNotNull() {
            assertNotNull(client.queries.getUser("user-with-posts").block().getGetUser().getPosts());
        }

        @Test
        void user_posts_has4Entry() {
            assertEquals(4, client.queries.getUser("user-with-posts").block().getGetUser().getPosts().size());
        }

        @Test
        void user_posts_firstEntry_isPostType() {
            User user = client.queries.getUser("user-with-posts").block().getGetUser();
            assertInstanceOf(Post.class, user.getPosts().get(0));
        }

        @Test
        void user_posts_firstEntry_hasCorrectId() {
            User user = client.queries.getUser("user-with-posts").block().getGetUser();
            assertEquals("post-10", user.getPosts().get(0).getId());
        }

        @Test
        void user_posts_firstEntry_hasCorrectTitle() {
            User user = client.queries.getUser("user-with-posts").block().getGetUser();
            assertEquals("Cyclic Post", user.getPosts().get(0).getTitle());
        }

        @Test
        void user_posts_author_isTruncatedUserType() {
            User user = client.queries.getUser("user-with-posts").block().getGetUser();
            assertInstanceOf(User.class,
                    user.getPosts().get(0).getAuthor());
        }

        @Test
        void user_posts_author_hasCorrectId() {
            User user = client.queries.getUser("user-with-posts").block().getGetUser();
            assertEquals("user-1", user.getPosts().get(0).getAuthor().getId());
        }

        @Test
        void user_posts_author_address_isNotNull() {
            User user = client.queries.getUser("user-with-posts").block().getGetUser();
            assertNotNull(user.getPosts().get(0).getAuthor().getAddress());
        }

        @Test
        void user_posts_author_address_city_isAccessible() {
            User user = client.queries.getUser("user-with-posts").block().getGetUser();
            assertEquals("Springfield",
                    user.getPosts().get(0).getAuthor().getAddress().getCity());
        }
    }

    // ── User with posts: null ─────────────────────────────────────────────────

    @Nested
    class UserNullPosts {

        @Test
        void user_posts_isNull_forAlice() {
            User user = client.queries.getUser("user-1").block().getGetUser();
            assertNull(user.getPosts());
        }
    }

    // ── Post → author (truncated user, cycle broken) ──────────────────────────

    @Nested
    class PostCycleAuthor {

        @Test
        void post_author_isTruncatedUserType() {
            assertInstanceOf(User.class,
                    client.queries.getPost("post-10").block().getGetPost().getAuthor());
        }

        @Test
        void post_author_hasCorrectId() {
            assertEquals("user-1",
                    client.queries.getPost("post-10").block().getGetPost().getAuthor().getId());
        }

        @Test
        void post_author_address_isNotNull() {
            assertNotNull(
                    client.queries.getPost("post-10").block().getGetPost().getAuthor().getAddress());
        }

        @Test
        void post_coAuthor_isNull() {
            assertNull(client.queries.getPost("post-10").block().getGetPost().getCoAuthor());
        }
    }

    // ── Post with coAuthor ────────────────────────────────────────────────────

    @Nested
    class PostWithCoAuthor {

        @Test
        void post_coAuthor_isNotNull() {
            assertNotNull(client.queries.getPost("post-2").block().getGetPost().getCoAuthor());
        }

        @Test
        void post_coAuthor_isTruncatedUserType() {
            assertInstanceOf(User.class,
                    client.queries.getPost("post-2").block().getGetPost().getCoAuthor());
        }

        @Test
        void post_coAuthor_hasCorrectId() {
            assertEquals("user-2",
                    client.queries.getPost("post-2").block().getGetPost().getCoAuthor().getId());
        }

        @Test
        void post_author_and_coAuthor_haveDifferentIds() {
            Post post = client.queries.getPost("post-2").block().getGetPost();
            assertNotEquals(post.getAuthor().getId(), post.getCoAuthor().getId());
        }
    }
}

File: integration_tests/kotlin_client_tests_real/client-app/src/test/kotlin/dev/graphlink/test/CycleTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.types.Post
import dev.graphlink.test.generated.types.User
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.newRealClient

class CycleTest {

    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        client = newRealClient()
    }

    // ── User → posts → post.author (cycle broken) ─────────────────────────────

    @Nested
    inner class UserWithPosts {

        @Test
        fun user_posts_isNotNull() = runTest {
            assertNotNull(client.queries.getUser("user-with-posts").getUser.posts)
        }

        @Test
        fun user_posts_hasOneEntry() = runTest {
            assertEquals(4, client.queries.getUser("user-with-posts").getUser.posts!!.size)
        }

        @Test
        fun user_posts_firstEntry_isPostType() = runTest {
            assertIs<Post>(client.queries.getUser("user-with-posts").getUser.posts!![0])
        }

        @Test
        fun user_posts_firstEntry_hasCorrectId() = runTest {
            assertEquals("post-10", client.queries.getUser("user-with-posts").getUser.posts!![0].id)
        }

        @Test
        fun user_posts_firstEntry_hasCorrectTitle() = runTest {
            assertEquals("Cyclic Post", client.queries.getUser("user-with-posts").getUser.posts!![0].title)
        }

        @Test
        fun user_posts_author_isTruncatedUserType() = runTest {
            assertIs<User>(
                client.queries.getUser("user-with-posts").getUser.posts!![0].author
            )
        }

        @Test
        fun user_posts_author_hasCorrectId() = runTest {
            assertEquals("user-1", client.queries.getUser("user-with-posts").getUser.posts!![0].author?.id)
        }

        @Test
        fun user_posts_author_address_isNotNull() = runTest {
            assertNotNull(client.queries.getUser("user-with-posts").getUser.posts!![0].author?.address)
        }

        @Test
        fun user_posts_author_address_city_isAccessible() = runTest {
            assertEquals(
                "Springfield",
                client.queries.getUser("user-with-posts").getUser.posts!![0].author?.address?.city,
            )
        }
    }

    // ── User with posts: null ─────────────────────────────────────────────────

    @Nested
    inner class UserNullPosts {

        @Test
        fun user_posts_isNull_whenNotPopulated() = runTest {
            val user: User = client.queries.getUser("user-1").getUser
            assertNull(user.posts)
        }
    }

    // ── Post → author (truncated user) ────────────────────────────────────────

    @Nested
    inner class PostCycleAuthor {

        @Test
        fun post_author_isTruncatedUserType() = runTest {
            assertIs<User>(
                client.queries.getPost("post-10").getPost.author
            )
        }

        @Test
        fun post_author_hasCorrectId() = runTest {
            assertEquals("user-1", client.queries.getPost("post-10").getPost.author?.id)
        }

        @Test
        fun post_author_address_isNotNull() = runTest {
            assertNotNull(client.queries.getPost("post-10").getPost.author?.address)
        }

        @Test
        fun post_coAuthor_isNull() = runTest {
            assertNull(client.queries.getPost("post-10").getPost.coAuthor)
        }
    }

    // ── Post with coAuthor ────────────────────────────────────────────────────

    @Nested
    inner class PostWithCoAuthor {

        @Test
        fun post_coAuthor_isNotNull() = runTest {
            assertNotNull(client.queries.getPost("post-2").getPost.coAuthor)
        }

        @Test
        fun post_coAuthor_isTruncatedUserType() = runTest {
            assertIs<User>(
                client.queries.getPost("post-2").getPost.coAuthor
            )
        }

        @Test
        fun post_coAuthor_hasCorrectId() = runTest {
            assertEquals("user-2", client.queries.getPost("post-2").getPost.coAuthor!!.id)
        }

        @Test
        fun post_author_and_coAuthor_haveDifferentIds() = runTest {
            val post: Post = client.queries.getPost("post-2").getPost
            assertNotEquals(post.author?.id, post.coAuthor!!.id)
        }
    }
}
